storage/src/main/scala/coop/rchain/storage/regex/Fsm.scala

```diff
@@ -0,0 +1,757 @@
+  /**
+    * In rare cases we need Fsm.AnythingElse to be the last symbol on the alphabet
+    */
+  private[this] def sortAlphabet(alphabet: Set[Char]): Seq[Char] =
+    if (alphabet.contains(anythingElse)) {
+      alphabet.toList.sortBy(sym =>
+        if (sym == anythingElse) Int.MaxValue else sym.toInt)
+    } else {
+      alphabet.toSeq
+    }
+
+  /**
+    * Given the above conditions and instructions, crawl a new unknown FSM,
+    * mapping its states, final states and transitions. Return the new FSM.
+    * This is a pretty powerful procedure which could potentially go on
+    * forever if you supply an evil version of follow().
+    */
+  private[regex] def crawl[T](alphabet: Set[Char],
+                               initial: T,
+                               isFinal: T => Boolean,
+                               follow: (T, Char) => Option[T]): Fsm = {
+    //actual type of the 'states' is deducted from 'initial' => ArrayBuffer[T]
+    val states = ArrayBuffer(initial)
+    val transitions = mutable.Map[Int, Map[Char, Int]]()
+
+    val sortedAlphabet = sortAlphabet(alphabet)
+    //we can't use any kind of range loop or iterators here, since collection is growing
+    @tailrec
+    def iterate(currentStateIdx : Int = 0, currentFinals : Set[Int] = Set()) : Set[Int] = {
+      val currentState = states(currentStateIdx)
+      //add to finals if needed
+      val nextFinals = if (isFinal(currentState)) {
+        currentFinals + currentStateIdx
+      } else {
+        currentFinals
+      }
+
+      val currentStateMap = sortedAlphabet.flatMap(symbol =>
+          follow(currentState, symbol).flatMap(nextState => {
+            val nextStateIdx = states.indexOf(nextState)
+            Some(symbol -> (if (nextStateIdx < 0) {
+              states += nextState
+              states.size - 1
+            } else {
+              nextStateIdx
+            }))
+          }))
+        .toMap
+
+      transitions += currentStateIdx -> currentStateMap
+
+      if (currentStateIdx + 1 < states.size)
+        iterate(currentStateIdx + 1, nextFinals)
+      else
+        nextFinals
+    }
+
+    val iterateFinals = iterate()
+    Fsm(alphabet, states.indices.toSet, 0, iterateFinals, transitions.toMap)
+  }
+
+  /**
+    * Crawl several FSMs in parallel, mapping the states of a larger meta-FSM.
+    * To determine whether a state in the larger FSM is final, pass all of the
+    * finality statuses (e.g. [True, False, False] to `test`.
+    */
+  def parallel(finalityTest: Seq[Boolean] => Boolean, fsms: Seq[Fsm]): Fsm = {
+    val alphabet = fsms.flatMap(_.alphabet).toSet
+
+    val initial = fsms.zipWithIndex.map{case (fsm, fsmIndex) => fsmIndex -> fsm.initialState}.toMap
+
+    def follow(currentState: GenMap[Int, Int], symbol: Char): Option[GenMap[Int, Int]] = {
+      val next = mutable.Map[Int, Int]()
+
+      fsms.zipWithIndex.foreach { case (fsm, fsmIdx) =>
+        val fsmState = currentState.get(fsmIdx)
+        if (fsmState.isDefined) {
```

Comment:
You can probably turn this whole if expression into a for expression and get rid of these `get` and `get.get`s.

Reply:
No get.get.get anymore, and even better - most of mutable.Sets have gone as well.